File: tests/test_inquiry.py
#!/usr/bin/env python3
"""Check the inquiry's labels and offline/activated modes without network calls."""
from html.parser import HTMLParser
from pathlib import Path
import re

ROOT = Path(__file__).resolve().parents[1]
PLACEHOLDER = 'AEGIS_WEB3FORMS_ACCESS_KEY_REQUIRED'
ENDPOINT = 'https://api.web3forms.com/submit'


class Inquiry(HTMLParser):
    def __init__(self, html):
        super().__init__()
        self.form = None
        self.fields = {}
        self.labels = set()
        self.in_form = False
        self.links = []
        self.scripts = []
        self.feed(html)

    def handle_starttag(self, tag, pairs):
        attrs = dict(pairs)
        if tag == 'form' and attrs.get('id') == 'project-inquiry':
            self.form = attrs
            self.in_form = True
        if tag == 'script' and attrs.get('src'):
            self.scripts.append(attrs['src'])
        if tag == 'a' and attrs.get('href'):
            self.links.append(attrs['href'])
        if not self.in_form:
            return
        if tag == 'label':
            self.labels.add(attrs.get('for'))
        if tag in ('input', 'select', 'textarea') and attrs.get('name'):
            self.fields[attrs['name']] = attrs

    def handle_endtag(self, tag):
        if tag == 'form':
            self.in_form = False


def check_markup(root=ROOT):
    html = (root / 'index.html').read_text()
    parsed = Inquiry(html)
    assert parsed.form is not None
    assert parsed.form['method'].lower() == 'post'
    online = parsed.form['action'] == ENDPOINT
    if online:
        assert 'enctype' not in parsed.form
        assert re.fullmatch(r'[0-9a-fA-F]{8}(?:-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}',
                            parsed.fields['access_key']['value'])
        assert parsed.fields['access_key']['value'] != PLACEHOLDER
        assert 'disabled' not in parsed.fields['access_key']
        assert 'disabled' not in parsed.fields['redirect']
        assert 'Submit your outline here' in html and 'Send inquiry' in html
        assert 'Email drafts go through your mail app' not in html
    else:
        assert parsed.form['enctype'] == 'text/plain'
        assert parsed.form['action'].startswith('mailto:admin@aegisequity.ca?')
        assert 'opens an email draft' in html and 'review and send it there' in html
        assert parsed.fields['access_key']['value'] == PLACEHOLDER
        assert 'disabled' in parsed.fields['access_key'] and 'disabled' in parsed.fields['redirect']
    assert 'inquiry.js' in parsed.scripts
    assert '#contact' in parsed.links and any(link.startswith('tel:') for link in parsed.links)
    assert any(link.startswith('mailto:admin@aegisequity.ca') for link in parsed.links)
    assert 'Web3Forms processes submissions' in html
    assert parsed.fields['redirect']['value'] == 'https://aegisequity.ca/thank-you.html'
    assert parsed.fields['botcheck']['type'] == 'checkbox'
    assert parsed.fields['botcheck']['tabindex'] == '-1'
    for name, required in (('name', True), ('email', True), ('phone', False),
                           ('project_type', True), ('message', True)):
        field = parsed.fields[name]
        assert field['id'] in parsed.labels, f"Missing label: {name}"
        assert ('required' in field) is required, f"Unexpected required state: {name}"
    assert parsed.fields['email']['type'] == 'email'
    assert parsed.fields['phone']['type'] == 'tel'
    assert parsed.fields['message']['maxlength'] == '1200'
    assert 'type="file"' not in html


if __name__ == '__main__':
    check_markup()
    print('PASS: inquiry form is labeled and correctly configured for its current mode.')

File: tests/test-inquiry.mjs
// Exercise inquiry.js without network access or a browser dependency.
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import vm from 'node:vm';

const script = readFileSync(new URL('../inquiry.js', import.meta.url), 'utf8');
const placeholder = 'AEGIS_WEB3FORMS_ACCESS_KEY_REQUIRED';
const endpoint = 'https://api.web3forms.com/submit';

function harness({ activated = true, valid = true, bot = false, fetchImpl } = {}) {
  const handlers = {};
  const status = { textContent: '', dataset: {}, focusCount: 0, focus() { this.focusCount++; } };
  const button = { disabled: false };
  const key = { disabled: !activated, value: activated ? 'TEST_AEGIS_ONLY_NOT_A_KEY' : placeholder };
  const form = {
    attributes: {},
    elements: { namedItem: name => ({ access_key: key, botcheck: { checked: bot } })[name] },
    getAttribute: name => name === 'action' ? (activated ? endpoint : 'mailto:admin@aegisequity.ca') : undefined,
    addEventListener: (event, handler) => { handlers[event] = handler; },
    querySelector: () => button,
    checkValidity: () => valid,
    reportValidityCount: 0,
    reportValidity() { this.reportValidityCount++; },
    resetCount: 0,
    reset() { this.resetCount++; },
    setAttribute(name, value) { this.attributes[name] = value; },
    removeAttribute(name) { delete this.attributes[name]; }
  };
  const calls = [];
  const context = {
    document: { getElementById: id => id === 'project-inquiry' ? form : status },
    FormData: class { constructor() { return [['access_key', key.value], ['name', 'Tester'], ['message', 'A small example project'], ['redirect', 'https://aegisequity.ca/thank-you.html']]; } },
    fetch: (...args) => { calls.push(args); return fetchImpl(...args); },
    AbortController,
    setTimeout,
    clearTimeout
  };
  vm.runInNewContext(script, context);
  return { handlers, status, button, form, calls, submit: () => {
    assert.ok(handlers.submit, 'Activated form registers a submit listener');
    let prevented = false;
    const done = handlers.submit({ preventDefault() { prevented = true; } });
    assert.ok(prevented, 'Activated form prevents native navigation');
    return done;
  } };
}

const dormant = harness({ activated: false, fetchImpl: () => { throw Error('must not fetch'); } });
assert.equal(dormant.handlers.submit, undefined, 'No-key form remains a native mailto draft');
assert.equal(dormant.calls.length, 0);

const pending = [];
const success = harness({ fetchImpl: () => new Promise(resolve => pending.push(resolve)) });
const first = success.submit();
assert.equal(success.button.disabled, true);
assert.equal(success.form.attributes['aria-busy'], 'true');
await success.submit();
assert.equal(success.calls.length, 1, 'Rapid duplicate submit starts only one request');
const [url, options] = success.calls[0];
assert.equal(url, endpoint);
assert.equal(options.method, 'POST');
assert.equal(options.headers['Content-Type'], 'application/json');
assert.equal(JSON.parse(options.body).redirect, undefined, 'No redirect sent with JSON fetch');
pending[0]({ ok: true, json: async () => ({ success: true }) });
await first;
assert.equal(success.form.resetCount, 1);
assert.equal(success.status.dataset.state, 'success');
assert.equal(success.form.attributes['aria-busy'], undefined);
assert.equal(success.button.disabled, true, 'Successful submission cannot be repeated');
await success.submit();
assert.equal(success.calls.length, 1);

for (const [label, response] of [
  ['HTTP failure', { ok: false, status: 400, json: async () => ({ success: false, body: { message: 'invalid key' } }) }],
  ['rate limit', { ok: false, status: 429, json: async () => ({ success: false, message: 'Too many requests' }) }],
  ['false success', { ok: true, status: 200, json: async () => ({ success: false }) }],
  ['invalid JSON', { ok: true, status: 200, json: async () => { throw Error('bad JSON'); } }]
]) {
  const h = harness({ fetchImpl: async () => response });
  await h.submit();
  assert.equal(h.form.resetCount, 0, `${label} must retain entries`);
  assert.equal(h.button.disabled, false, `${label} must permit retry`);
  assert.equal(h.status.dataset.state, 'error', `${label} must report error`);
  assert.ok(h.status.focusCount, `${label} must focus status`);
}
const offline = harness({ fetchImpl: async () => { throw Error('network down'); } });
await offline.submit();
assert.equal(offline.status.dataset.state, 'error');
assert.equal(offline.form.resetCount, 0);
assert.match(offline.status.textContent, /could not confirm delivery/);

const bot = harness({ bot: true, fetchImpl: () => { throw Error('must not fetch'); } });
await bot.submit();
assert.equal(bot.calls.length, 0);
assert.equal(bot.status.dataset.state, 'error');

const invalid = harness({ valid: false, fetchImpl: () => { throw Error('must not fetch'); } });
await invalid.submit();
assert.equal(invalid.form.reportValidityCount, 1);
assert.equal(invalid.calls.length, 0);

console.log('PASS: dormant fallback, validation, honeypot, duplicate guard, success and error responses.');
